[Spring 지하철 노선도 - 3단계] 정(정진우) 미션 제출합니다.

bugoverdose opened the pull request.

bugoverdose committed:
docs(README): add step3 feature list

bugoverdose committed:
feat(db): add section table to all the schemas

- add unique key constraint on foreign id combination

bugoverdose committed:
feat(domain): implement Section entity

bugoverdose committed:
feat(dao): implement SectionDao

- add tests

bugoverdose committed:
refactor(test): remove duplicate code for test db setup & cleansing

- TODO: should fix test fixture for services

bugoverdose committed:
refactor(domain): change domain package to entity and change class names

bugoverdose committed:
refactor(ui): remove unnecessary produces option

bugoverdose committed:
refactor(test): overall cleansing

bugoverdose committed:
feat(dto): separate request dto for POST and PUT

bugoverdose committed:
feat(service): implement new LineService2 and LineResponse2 dto

- implement save method for saving new Line
- add test cases and new fixture

bugoverdose committed:
feat(dao): implement findAllByIds method for stations

bugoverdose committed:
refactor(service): cleanse validation logic before saving new line

bugoverdose committed:
feat(service): implement find method at LineService2

- view Line by id with all the stations
- add test cases

bugoverdose committed:
feat(service): implement delete method at LineService2

- delete line and all the sections referencing the line
- add test cases

bugoverdose committed:
feat(dao): implement basic implement findAll method for sections

- add test cases
- change test fixture and fix broken tests

bugoverdose committed:
feat(dao): change save logic to return void

- because it's unnecessary
- fix test case

bugoverdose committed:
feat(dao): implement section queries that return SectionViewEntity

- fix test case
- sql tuning by joining station and section

bugoverdose committed:
refactor(service): cleanse find logic

- replace legacy findAllByLineId with new dao method

bugoverdose committed:
feat(dao): implement FullStationDao

- findAll method joins line entity for each station entity

bugoverdose committed:
feat(service): implement finding all lines at LineService2

- findAll method returns all the lines sorted by lineId ascending order
- add test fixture for LineService2 test

bugoverdose committed:
refactor(test): separate test db configuration for /lines and /stations

bugoverdose committed:
feat(controller): replace legacy service with new LineService2

- fix broken tests
- use old LineRequest as update request dto

bugoverdose committed:
refactor(subway): replace classes and remove all unused logic

bugoverdose committed:
feat(subway): implement validation for creating new line

- add test cases

bugoverdose committed:
docs(README): update feature list

bugoverdose committed:
feat(dao): integrate multiple delete logic to single delete query

bugoverdose committed:
refactor(dao): implement SectionViewDao to separate logic for view

- refactor usages and test cases

bugoverdose committed:
feat(dao): re-implement findAllByLineId with default SectionEntity format

bugoverdose committed:
test(study): check logic for converting edges to sorted nodes

bugoverdose committed:
feat(subway): sort stations in ascending order on GET /lines/:id

- replace Sections with SectionViews
- sort from upStation to downStation
- change test cases to check sorting

bugoverdose committed:
refactor(dao): select only necessary columns for SectionView

- remove unnecessary fields on SectionViewEntity
- fix test case

bugoverdose committed:
feat(domain): implement basic Section domains

bugoverdose committed:
refactor(SectionViews): use section domains for fields to cleanse logic

bugoverdose committed:
refactor(domain): move SectionViews to domain package

bugoverdose committed:
feat(domain): implement Sections and SectionDistanceMap

- first class collection for all sections in a line

bugoverdose committed:
feat(service): implement SectionService

- implement deleting section from line by stationId
- add test cases

bugoverdose committed:
refactor(test): cleanse service tests

- use service test db instead of dao test db
- remove unnecessary annotations and fields

bugoverdose committed:
refactor(test): cleanse acceptance tests

- cleanse setup and cleanse logic

bugoverdose committed:
feat(controller): implement controller for DELETE /lines/:id/sections?stationId={stationId}

- add test cases

bugoverdose committed:
docs(README): update feature list

bugoverdose committed:
feat(dao): implement update methods for section

bugoverdose committed:
feat(service): implement saving section

- add test cases

bugoverdose committed:
feat(controller): implement controller for POST /lines/:id/sections

- add test cases

bugoverdose committed:
refactor(test): cleanse all test setup

bugoverdose committed:
docs(README): update feature list

bugoverdose committed:
refactor(domain): cleanse verbose logic for saving new section

bugoverdose committed:
refactor(domain): rename Section to SectionStationInfo

bugoverdose committed:
refactor(domain): rename Section2 to Section

jjj0611 commented on src/main/java/wooteco/subway/entity/FullStationEntity.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.util.Objects;
+
+public class FullStationEntity {
```
FullStationEntity의 역할에 대해 한 번 설명해주시겠어요 정?

bugoverdose replied:
음. "구간에 등록된 지하철역 정보"라고 요약할 수 있겠네요. 개별 지하철역에 대해 자신이 구간으로 등록된 노선에 대한 정보를 함께 묶은 엔티티입니다.

SectionEntity의 경우 개별 구간에 대해 두 개의 지하철역이 함께 묶였는데, FullStationEntity는 구간이 아니라 개별 지하철역에 초점을 맞추고 있습니다.

예를 들어 노선1과 노선2에 구간으로 등록된 지하철역1이 있는 경우, `FullStationEntity{LineEntity(id=1), StationEntity(id=1)}`과 `FullStationEntity{LineEntity(id=2), StationEntity(id=1)}`이 조회됩니다.

cf) 현재는 `RegisteredStationEntity` & `RegisteredStationDao`로 이름 수정했습니다.

jjj0611 commented on src/main/java/wooteco/subway/entity/SectionViewEntity.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Objects;
+import wooteco.subway.domain.SectionStationInfo;
+
+public class SectionViewEntity {
```
SectionViewEntity에 대한 설명도 한 번 해주세요 :)

bugoverdose replied:
기존 `SectionEntity`의 경우 section 테이블에 있는 지하철역들의 id값만을 지닙니다.

노선 단건 조회(`GET /lines/:id`)의 경우 id에 대응되는 실제 지하철역들의 name 정보도 필요해서, StationEntity(id, name)으로 구성된 SectionViewEntity를 만들었습니다.

jjj0611 commented on src/main/java/wooteco/subway/entity/FullStationEntity.java:
```diff
@@ -0,0 +1,52 @@
+
+import java.util.Objects;
+
+public class FullStationEntity {
```
더불어서 FullStation이란 네이밍이 적절한지도 한 번 고민 부탁드려요 :)

bugoverdose replied:
현재 이걸 `RegisteredStationEntity`로 수정한 상황인데요.

이런식으로 하면 `registered_station`이라는 테이블이 존재하는 것처럼 느껴지지 않은가?라는 생각이 들어서 고민했었습니다.

"station 테이블의 지하철역 정보를 기반으로 JOIN 등의 방법으로 약간의 데이터를 추가해서 DB에서 가져온 데이터다"라는 느낌을 살릴 방법이 잘 안떠오르네요. (e.g., `StationWithLineEntity`?)

jjj0611 commented on src/main/java/wooteco/subway/entity/SectionViewEntity.java:
```diff
@@ -0,0 +1,54 @@
+import java.util.Objects;
+import wooteco.subway.domain.SectionStationInfo;
+
+public class SectionViewEntity {
```
SectionView와 Section을 나눴을 때 어떤 효용이 있었나요??

bugoverdose replied:
도메인을 말씀하시는 거라면, 하나는 저장/생성 관련 로직을 처리하고, 다른 하나는 조회와 관련된 로직을 처리하는데 사용할 수 있었습니다.

`Section`의 경우 `Sections`를 통해 특정 노선을 재구성하여, 해당 구조에 새로운 구간을 저장할 수 있는지, 특정 구간을 제거할 수 있는지를 판별하는 로직과 관련있습니다.

`SectionView`의 경우 `SectionViews`에서 노선을 구성하는 구간들을 상행종점부터 하행종점까지 정렬하는 로직과 관련이 있습니다. 특정 노선을 조회할 때 사용됩니다.

bugoverdose replied:
엔티티를 말씀하시는거라면 SectionEntity는 section 테이블의 정보를 그대로 가져오고, SectionViewEntity는 section 테이블의 정보를 토대로 station 정보를 가져오는 기능을 수행하네요.

bugoverdose replied:
흠. 둘을 하나로 합치는게 도메인 구조를 깔끔하게 만들 것 같긴 하네요.

jjj0611 replied:
저도 지금 상태에선 그렇다고 생각해요.
둘 중의 한 방향으로 통일을 했다가 필요성을 느끼면 그 때 분리해봤어도 좋을 것 같아요 :)

bugoverdose replied:
넵 (1) 단순하게 시작해서 (2) 확실한 구조를 만들고, (3) 그 다음에 변형을 조금씩 시도하는게 맞는 것 같다는 생각이 확실히 드네요.

jjj0611 commented on src/main/java/wooteco/subway/domain/Section.java:
```diff
@@ -0,0 +1,56 @@
+
+    private static final String SAME_STATION_INPUT_EXCEPTION = "서로 다른 두 개의 역을 입력해야 합니다.";
+
+    private final Long upStationId;
+    private final Long downStationId;
```
section은 도메인 객체 아닌가요?
Station 객체를 사용해도 되지 않나요?

bugoverdose replied:
section을 구성할 때 (사용하지 않을 예정이더라도) 일단 지하철역의 name까지 가져오라는 뜻이 맞나요?

jjj0611 replied:
정 조금 바꿔서 
만약 db 등록 기능이 없는(애플리케이션이 꺼지면 데이터가 초기화 되어도 됨) 애플리케이션이었다면,
즉 레벨1과 같은 상태였다면 어떻게 도메인 구조를 짜셨을 것 같나요?

jjj0611 replied:
음,, 다른 코멘트에서도 남겼지만,
애플리케이션이 종료되면 데이터가 날아가도 되는 레벨1의 DB가 없는 상태라고 생각했을 때
정은 도메인을 어떻게 구성할 것 같나요?

DB가 연결되었다고 해서 도메인이 바뀌어야 하는 것은 아닙니다

jjj0611 commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,87 @@
+    public void save(Long lineId, CreateSectionRequest sectionRequest) {
+        Section newSection = sectionRequest.toSection();
+        Sections sections = Sections.of(findValidSections(lineId));
+        validateExistingStations(newSection);
+        sections.validateSingleRegisteredStation(newSection);
+
+        if (!sections.isEndSection(newSection)) {
+            SectionEntity updatedPreviousSection = sections.toUpdatedSection(lineId, newSection);
+            boolean containsRegisteredUpStation = sections.hasRegisteredUpStation(newSection);
+            updatePreviousSection(updatedPreviousSection, containsRegisteredUpStation);
+        }
```
section 등록에 대한 로직들이 모두 서비스에 들어갔네요.
```
sections.add(section)
secstionDao.save(sections);
```
로 간단하게 풀 순 없을까요?

bugoverdose replied:
저도 서비스 로직이 이런 구조가 되는게 가장 이상적일 것 같다고 생각은 했는데, 

모든 구간 정보를 다 삭제하고, 전부 다시 저장하는 것에 거부감을 느껴서 지금처럼 해봤었어요.

현재 sections와 기대되는 sections를 비교하고, "삭제해야 하는 정보"와 "저장해야 하는 정보"를 서비스에서 반환받을 방법은 없나 고민이 됩니다.

```java
secstionDao.delete(newSections.findDeletedSections(oldSections));
secstionDao.save(newSections.findNewSections(oldSections));
```

이런 느낌처럼요

jjj0611 replied:
오 이렇게 될 수 있어도 너무 좋죠.
한 번 고민해봐주세요 :)

jjj0611 replied:
미션 진행하시면서 계속 고민하시고 적용해보시면 좋을 것 같아요 :)

jjj0611 commented on src/test/resources/section_test_fixture.sql:
```diff
@@ -0,0 +1,25 @@
+ALTER TABLE line ALTER COLUMN id RESTART WITH 1;
+ALTER TABLE section ALTER COLUMN id RESTART WITH 1;
+
+INSERT INTO station(id, name)
+VALUES (1, '강남역'),
+       (2, '선릉역'),
+       (3, '잠실역'),
+       (4, '청계산입구역'),
+       (5, '한티역');
+
+INSERT INTO line(id, name, color)
+VALUES (1, '1호선', '노란색'),
+       (2, '신분당선', '빨간색'),
+       (3, '2호선', '초록색');
+
+INSERT INTO section(id, line_id, up_station_id, down_station_id, distance)
+VALUES (1, 1, 3, 1, 5),
+       (2, 2, 2, 3, 5),
+       (3, 2, 1, 2, 5),
+       (4, 3, 1, 3, 7);
```
절충안에 대해 고민해보셨나요?
fixture들이 n개가 되면서 더 보기가 어려워졌어요.
정이 혼자 개발하는 애플리케이션이라면 문제가 되지 않겠지만,
다른 사람과 함께 개발하는 환경에서도 그럴지 고민해주세요 :)

bugoverdose replied:
음 도메인쪽이 정리되면 `TestFixtureUtil.saveLines(~)`와 같은 메서드를 제공하는 유틸성 클래스를 test 패키지쪽에 만들어보는걸 생각하고 있었습니다.

jjj0611 replied:
지금처럼 n개의 test fixture를 위한 데이터 생성이 sql 파일로 들어가게 되면,
개발자들이 어떤 상황에 어떤 픽스쳐 sql을 사용해야할지를 알아야 하고,
이게 더 복잡함을 만들거에요.

이런 경우는 오히려 각각 테스트마다 필요한 데이터를 코드 상에 위치시켜야
코드만 보고도 테스트를 파악할 수 있을 겁니다 :)

jjj0611 commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,108 @@
+import java.util.stream.Collectors;
+import wooteco.subway.entity.SectionEntity;
+
+public class Sections {
```
현재 domain 자체에 대한 테스트는 한 개도 보이지 않네요.
Service test로 모든걸 대체하신건가요??

bugoverdose replied:
도메인 구조를 많이 바꿔야 할 것 같아서 리뷰를 받고, 리팩토링을 한 후에, 테스트를 작성하는것을 생각하고 있었습니다.

jjj0611 replied:
테스트가 있는 상태에서 리팩토링을 해야 더 안정적인 리팩토링이 되지 않을까요?

bugoverdose committed:
refactor(entity): rename FullStationEntity to RegisteredStationEntity

- change dao name
- fix usages

jjj0611 commented on src/main/java/wooteco/subway/domain/Section.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.List;
+import wooteco.subway.entity.SectionEntity;
+
+public class Section {
```
전체적으로 domain 패키지에 대한 역할 정리를 한 번 글로 풀어주시겠어요 정?
Section과 Sections의 관계 / Sections와 SectionDistanceMap, SectionStationMap의 관계, SectionDistanceMap이 하고 있는 역할과 StationStationMap이 하고 있는 역할 등 domain 패키지에 있는 모든 클래스에 대한 역할 정리를 글로 한 번 풀어봐주세요 :)

bugoverdose replied:
답변 다는걸 또 잊었네요. 현재 전부 없어진 객체들이라 잘 기억이 안나는데, 일단 Map들은 전부 파이썬으로 알고리즘 문제 푸는 느낌으로 정렬 작업하려고 구현했던거였어요. 새로 만든 도메인들의 생성자를 보면 Map으로 수행했던 로직으로 정렬 작업을 수행하고 있는게 좀 남아있네요.

현재 도메인들은 테스트 코드로 최대한 로직들 설명하고 있으니, 저번보다는 덜 난해하지 않을까 싶어요.

jjj0611 commented on src/main/java/wooteco/subway/domain/StationEntityMap.java:
```diff
@@ -0,0 +1,53 @@
+import wooteco.subway.entity.SectionViewEntity;
+import wooteco.subway.entity.StationEntity;
+
+public class StationEntityMap {
```
전반적으로 domain과 entity를 나눈 것 같은데,
id를 key로 하는 StationEntity의 map에 대한 일급 컬렉션은 도메인 패키지에 존재하네요.
특별한 이유가 있나요?

bugoverdose replied:
원래는 StationEntity말고 Station(id, name)라는 도메인을 만들고, StationMap으로 만들어야 하나 고민했는데, 도메인이 id라는 값을 지니는게 맞나? 싶어서 일단은 내버려뒀어요.

Station이라는 도메인을 별도로 만들고는 싶은데 어떻게 만들면 좋을지 모르겠어서 리팩토링이 덜 된 상태네요.

jjj0611 replied:
엇 이 부분은 domain과 entity라는 클래스들의 패키지 정리가 잘 안되었다는 의미로 남긴 코멘트였어요 :)

bugoverdose committed:
feat(RegisteredStationDao): implement findAllByStationId

- select list of registered station data
- add test case

jjj0611 commented on src/main/java/wooteco/subway/service/SectionService.java:
```diff
@@ -0,0 +1,87 @@
+
+    @Transactional
+    public void delete(Long lineId, Long stationId) {
+        Sections sections = Sections.of(findValidSections(lineId));
+        sections.validateRegisteredStation(stationId);
+        validateNotLastSection(sections);
+
+        sectionDao.deleteAllByLineIdAndStationId(lineId, stationId);
+        if (sections.isMiddleStation(stationId)) {
+            sectionDao.save(sections.toConnectedSectionBetween(lineId, stationId));
+        }
```
```
sections.deleteBy(stationId);
sectionDao.save(sections);
```
와 같이는 어려울까요?

여기도 도메인 로직이 서비스 레이어에 펼쳐져 있는 것 같아요 :)

bugoverdose replied:
음. 조건문 하나 정도는 서비스에 있어도 괜찮지 않나 싶었는데, 포비가 말한 서비스를 최대한 얇게 만들어야 한다는게 이런 의미였나 싶네요.

jjj0611 replied:
네 서비스레이어에 대해서 좀 더 공부해보면 어떨까요? :)

jjj0611 commented on src/main/java/wooteco/subway/entity/SectionEntity.java:
```diff
@@ -0,0 +1,94 @@
+
+public class SectionEntity {
+
+    private static final int MIN_DISTANCE = 1;
```
이런건 domain로직일까요? entity 로직일까요?

전반적으로 domain과 entity의 역할 설정이 불분명한 것 같아요.
각 개념에 대해 학습해보시면 어떨까요??

bugoverdose replied:
음 비즈니스 로직이 아니라, DB에 저장되는 값에 대한 validation이니 DB에 가장 가까운 엔티티에 넣어야하지 않나 싶었는데 아닌가요?

bugoverdose replied:
일단은 이렇게 알고 있습니다.
- 엔티티는 DB의 테이블 대응되는 객체.
- 도메인은 문제해결에 필요한 로직을 담당하는 객체.

jjj0611 replied:
최소 거리가 0보다 커야한다는 것은 DB에 저장될 때의 문제(DB 관점의 로직)가 아닌 도메인 관점의 로직(비즈니스 로직)이지 않을까요?
다시 한 번 말하지만, DB에 저장되어야 한다는 요구사항이 없는 채로 이 애플리케이션을 만든다고 할 때,
station과 station간 거리가 0보다 커야한다는 로직은 어디에 위치하게 될까요?

jjj0611 reviewed (changes requested):
안녕하세요 정. 리뷰어 제이입니다!

3단계 구현하신 코드 잘 봤습니다.

코멘트 남겨드렸으니 확인 부탁드려요.
코멘트에도 남겼지만, 전반적으로 domain의 역할과 entity의 역할이 다소 모호한 것 같아요.

코멘트 확인해보시고 이해 안가거나 어려운 점이 있으면 코멘트나 DM 부탁드려요.

bugoverdose committed:
feat(StationService): prevent deleting registered stations

- add validation with RegisteredStationDao
- add and fix test cases

bugoverdose committed:
feat(test): check 400 response on trying to delete registered station

- add and cleanse test cases

bugoverdose committed:
docs(README): update feature list

bugoverdose committed:
feat(dao): implement new SectionDao and SectionEntity

- map StationEntities to sectionEntity

bugoverdose committed:
feat(domain): implement new SectionViews and Station with new SectionEntity

bugoverdose committed:
refactor(service): replace SectionViewDao with SectionDao2

- replace StationEntity to Station domain

bugoverdose committed:
refactor(service): replace SectionDao with SectionDao2

- implement basic Line domain

bugoverdose committed:
refactor(service): cleanse toLineResponse logic

bugoverdose committed:
refactor(LineService): cleanse findAll logic with Lines domain

bugoverdose committed:
refactor(domain): separate domains to sub-packages

bugoverdose committed:
feat(domain): implement Section

bugoverdose committed:
refactor(SectionEntity): remove validation logic to Section domain

- fix use case

bugoverdose committed:
refactor(domain): replace entities with Section domain

bugoverdose committed:
refactor(domain): cleanse domains for sorting stations

bugoverdose committed:
refactor(domain): rename SectionView to Stations

- remove legacy StationMap

bugoverdose committed:
feat(domain): implement Sections

- sort sections from upperEndStation to lowerEndStation
 - delete by station
 - compare and extract new/deleted sections

bugoverdose committed:
feat(service): implement SectionService2 with new domains

- implement deleteSection
- add test cases

bugoverdose committed:
feat(domain): implement save at Section2

bugoverdose committed:
feat(service): implement save at SectionService2

- add test cases

bugoverdose committed:
refactor(service): replace legacy SectionService with new service

bugoverdose committed:
test(dao): add all tests for SectionDao2

bugoverdose committed:
refactor(subway): replace legacy services, domains, entities

- remove unused code

bugoverdose committed:
refactor(domain): cleanse Sections logic

- implement SectionsFactory for creating Sections
- remove unused classes
- overall cleansing

bugoverdose committed:
refactor(domain): cleanse entity and domain logic

bugoverdose committed:
refactor(ui): validate request dto

- add spring validation gradle dependency
- remove null/blank input validation at domain
- add test cases

bugoverdose committed:
test(utils): implement TestFixtureManager for saving fixtures

bugoverdose committed:
refactor(test): setup dao test fixtures with TestFixtureManager

- setup only necessary fixture for each test
- remove unnecessary sql file

bugoverdose committed:
refactor(test): setup assurance test fixtures with TestFixtureManager

- setup only necessary fixture for each test
- remove unnecessary sql file

bugoverdose committed:
refactor(test): setup service test fixtures with TestFixtureManager

- setup only necessary fixture for each test
- remove unnecessary sql files

bugoverdose committed:
refactor(exception): cleanse
NotFoundException usages

- implement ExceptionType enum
 - remove duplicate exception message constants for station and line not found exceptions

bugoverdose committed:
refactor(subway): cleanse domains, entities and its usages

- remove toEntity and all dependency on entity package from domain package

bugoverdose committed:
test(ui): add minor test case

 - check status code 404 response on deleting section with non-existing subway

bugoverdose committed:
test(SectionsFactory): add test cases on generate

- add validation for section connection

bugoverdose committed:
feat(dao): implement RegisteredSectionDao

- implement findAll method and entity
- add test case

bugoverdose committed:
feat(domain): implement Lines, Line, RegisteredSection domains

- Lines returns sorted lines with sorted sections

bugoverdose committed:
feat(SectionDao): implement findAllByStationId

- add test case

bugoverdose committed:
refactor(service): replace all RegisteredStation logic

- replace RegisteredStationDao with other Daos
- sort all sections for each line on GET /lines
- update test case
- remove all unused files

bugoverdose committed:
refactor(domain): rename Line domains

- rename old Line to LineInfo
- rename new Line2 to Line

bugoverdose committed:
feat(RegisteredSectionDao): implement findAllByLineId method

- add test case

bugoverdose committed:
refactor(subway): cleanse finding single line logic

- cleanse service and domains

bugoverdose committed:
feat(domain): add more validation logic at Line

- add test cases

bugoverdose committed:
test(domain): add tests for Lines

bugoverdose committed:
refactor(subway): rename RegisteredSection to LineSection

- because each LineSection connects one LineEntity and one SectionEntity

bugoverdose committed:
test(domain): add tests for Section domain

- implement minor validation logic

bugoverdose committed:
feat(domain): implement Sections2

- sort section list on construction
- handle all simple first class collection logic

bugoverdose committed:
refactor(domain): rename Sections to SectionsManager and cleanse logic

- add Section2 for sorting and handling collection
- cleanse Section2 and SectionManager usage
- remove unnecessary SectionsFactory

bugoverdose committed:
refactor(domain): rename Sections2 to Sections

bugoverdose committed:
test(domain): add all test cases for SectionsManager

bugoverdose committed:
test(domain): add all test cases for Sections

- minor refactoring

bugoverdose committed:
refactor(service): rollback find line logic

- find single line with lineDao and sectionDao

bugoverdose committed:
feat(dao): implement findAll methods for Line and Section

- add test cases

bugoverdose committed:
feat(domain): implement domains for finding all Lines

- implement Section2 and Lines2

bugoverdose committed:
refactor(service): replace Lines with Lines2 for findAll

- remove LineSectionDao

bugoverdose committed:
refactor(domain): replace old Section, Lines with new domains

- fix use cases
- remove unused classes

bugoverdose committed:
test(domain): add tests on Section

bugoverdose committed:
feat(domain): add validation logic at Lines

- throw exception on lineInfos and sections none-match

bugoverdose committed:
refactor(subway): minor cleansing

bugoverdose committed:
feat(repository): implement SubwayRepository and StationRepository

- handle all dao and entity logic

bugoverdose committed:
refactor(service): replace daos with repositories

- remove entities from service (only use domains)

bugoverdose committed:
test(repository): add tests for StationRepository

bugoverdose committed:
test(repository): add tests for SubwayRepository

jjj0611 reviewed (approved):
안녕하세요 정. 리뷰어 제이입니다!

질문 남겨주신 코멘트에 답글 달았으니 확인 한 번 부탁드려요.
조금 덧붙이면
domain과 entity, 비즈니스 로직과 DB 연결에 대해 좀 더 공부해보면 어떨까요?

DB 연결이 아예 없다고 생각하고, 도메인 객체 간 일을 할 수 있도록 구성해보는 것이
공부하는데 조금 도움이 될 수도 있을 것 같아요.

일단 다음 미션 진행을 위해 코드는 여기서 머지할게요!
미션 진행하시면서 계속 고민해보시면 좋을 것  같아요 :)

지하철 노선도 미션 어려우셨을텐데 고생 많으셨습니다!
